x-pack/plugins/fleet/server/services/agents/actions.ts

```diff
@@ -234,61 +234,111 @@ export async function getUnenrollAgentActions(
+        if (action && action.agents && action.action_id && action.type === 'UPGRADE') {
+          acc.push(action);
+        }
+        return acc;
+      }, []);
+    return upgradeActions;
+  };
 
   const cancelActionId = uuidv4();
   const now = new Date().toISOString();
-  for (const hit of res.hits.hits) {
-    if (!hit._source || !hit._source.agents || !hit._source.action_id) {
-      continue;
-    }
-    if (hit._source.type === 'UPGRADE') {
-      const errors = {};
-      await bulkUpdateAgents(
-        esClient,
-        hit._source.agents.map((agentId) => ({
-          agentId,
-          data: {
-            upgraded_at: null,
-            upgrade_started_at: null,
-          },
-        })),
-        errors
-      );
-      if (Object.keys(errors).length > 0) {
-        appContextService
-          .getLogger()
-          .debug(`Errors while bulk updating agents for cancel action: ${JSON.stringify(errors)}`);
-      }
-    }
+
+  const cancelledActions: Array<{ agents: string[] }> = [];
+
+  const createAction = async (action: FleetServerAgentAction) => {
     await createAgentAction(esClient, {
       id: cancelActionId,
       type: 'CANCEL',
-      agents: hit._source.agents,
+      agents: action.agents!,
       data: {
-        target_id: hit._source.action_id,
+        target_id: action.action_id,
       },
       created_at: now,
-      expiration: hit._source.expiration,
+      expiration: action.expiration,
     });
+    cancelledActions.push({
+      agents: action.agents!,
+    });
+  };
+
+  let upgradeActions = await getUpgradeActions();
+  for (const action of upgradeActions) {
+    await createAction(action);
+  }
+
+  const updateAgentsToHealthy = async (action: FleetServerAgentAction) => {
+    appContextService
+      .getLogger()
+      .info(
+        `Moving back ${
+          action.agents!.length
+        } agents from updating to healthy state after cancel upgrade`
+      );
+    const errors = {};
+    await bulkUpdateAgents(
+      esClient,
+      action.agents!.map((agentId: string) => ({
+        agentId,
+        data: {
+          upgraded_at: null,
+          upgrade_started_at: null,
+        },
+      })),
+      errors
+    );
+    if (Object.keys(errors).length > 0) {
+      appContextService
+        .getLogger()
+        .info(`Errors while bulk updating agents for cancel action: ${JSON.stringify(errors)}`);
+    }
+  };
+
+  for (const action of upgradeActions) {
+    await updateAgentsToHealthy(action);
+  }
+
+  // find missing batch
+  upgradeActions = await getUpgradeActions();
+  if (cancelledActions.length < upgradeActions.length) {
+    const missingBatches = upgradeActions.filter(
+      (upgradeAction) =>
+        !cancelledActions.some(
+          (cancelled) => upgradeAction.agents && cancelled.agents[0] === upgradeAction.agents[0]
+        )
+    );
+    appContextService.getLogger().info(`missing batches to cancel: ${missingBatches.length}`);
```

Comment:
Is `info` the right log level here?

Reply:
will change it back to debug, used info in testing as I couldn't change the log level in qa env